api/adaptive_load/adaptive_load.proto

```diff
@@ -0,0 +1,68 @@
+syntax = "proto3";
+
+package nighthawk.adaptive_load;
+
+import "api/adaptive_load/benchmark_result.proto";
+import "api/adaptive_load/metrics_plugin.proto";
+import "api/adaptive_load/metric_spec.proto";
+import "api/adaptive_load/step_controller.proto";
+import "api/client/options.proto";
+import "google/protobuf/duration.proto";
+import "google/rpc/status.proto";
+import "validate/validate.proto";
+
+// Parameters describing the adjusting and testing stages of an adaptive load
+// session, which consists of a series of Nighthawk benchmarks probing for
+// the optimal load on the system, followed by a longer benchmark to validate
+// the values. Load adjustments are calculated by the selected StepController
+// plugin. Metrics can come from Nighthawk stats and counters or
+// platform-specific data sources via MetricsPlugins.
+message AdaptiveLoadSessionSpec {
+  // Settings for MetricsPlugins that obtain metrics from outside sources.
+  // An entry is required for every plugin referred to by metric_thresholds,
+  // other than the "builtin" plugin.
+  repeated MetricsPluginConfig metrics_plugin_configs = 1;
+  // Metrics and thresholds that determine load adjustments.
+  repeated MetricSpecWithThreshold metric_thresholds = 2 [(validate.rules).repeated .min_items = 1];
+  // Metrics that are collected and included in the output but not taken into
+  // account when adjusting the load.
+  repeated MetricSpec informational_metric_specs = 3;
+  // A proto describing Nighthawk Service traffic. See
+  // https://github.com/envoyproxy/nighthawk/blob/master/api/client/options.proto
+  //
+  // The adaptive load controller will return an error if the |duration| or
+  // |open_loop| fields are set within |nighthawk_traffic_options|. The
+  // controller will also be configured to overwrite at least one of the
+  // numerical fields during the search, such as requests_per_second, so any
+  // value of those fields specified here will be ignored.
+  //
+  // All other fields in |nighthawk_traffic_options| are passed through to the
+  // Nighthawk Service.
+  //
+  // Note that |concurrency| in |nighthawk_traffic_options| multiplies the total
+  // RPS sent.
+  nighthawk.client.CommandLineOptions nighthawk_traffic_template = 4;
+  // The duration of each short benchmark during the adjusting stage. Must be
+  // set. Reasonable value: 10 seconds.
+  google.protobuf.Duration measuring_period = 5 [(validate.rules).duration.gt.seconds = 0];
```

Comment:
just a random thought -- we're also able to terminate NH benchmarks on other conditions then a max duration, e.g. when a NH counter hits a threshold. I can't see direct value in that in the context of this PR at this time -- but I haven't spend time trying to see it either. So I'm just thinking out loud: maybe it is worth considering not pinning step termination to a duration, but leaving an escape latch in here for if someone ever would like to terminate steps based on other premises? Or maybe no conditional at all, and leave it up to an external sigterm? 
This also brings another thought: do we need to allow handling of error conditions, like should it be weighed somehow if there's a certain 5xx replies observed? (In the current code base these concepts are sort of similar/related in the way they are implemented, e.g. there's termination predicates, and failure predicates, hence me associating these things in a single comment). 
Possibly this can be handled as-is, but I'm just curious if failing to find what we're looking for, based on conditions other then the convergence deadline, is also an option, and what that would look like concretely in configuration.

Reply:
That's very interesting. I think we can actually fully support this as-is:

You can put any termination predicates in nighthawk_traffic_template so that it doesn't keep going for the full 10 second measuring period after causing fatal 5xxs.

If you anticipate 5xxs, you would also want to add a threshold on a builtin metric for something like success-rate, which is 2xx count / total count, so that the next RPS will be reduced on the basis of causing 5xx, even if latency still looked ok. We might recommend always including a threshold for success-rate.

I already implemented success-rate, but maybe there are other metrics that could be derived from NH counters. This code is coming a few PRs from now.

Reply:
Sounds great